### DISCUSSION TITLE
wdio 7.x how to extend browser.config types

### DISCUSSION THREAD

**Discussion by tzurp** [webdriverio/webdriverio#6739] [Q&A] [answered]
In wdio 6 we used the following to add custom config types:

```
declare module WebdriverIO {
    interface Config {
        customProp1: string;
        customProp2: string;
        }
}
```
in wdio 7 it is not working and also the following didn't work:
```
declare global {
    namespace WebdriverIO {
    interface Config {
        customProp1: string;
        customProp2: string;
        }
}
```
How should it be done with version 7?

**Comment by christian-bromann** [comment, marked as answer]
@tzurp it is currently not possible to extend the config type. Can you elaborate on your use case?

**Reply by tzurp** [reply]
@christian-bromann, we are using custom configs to define parameters like url paths, port numbers etc. that are taken from the wdio.conf files.

**Reply by christian-bromann** [reply]
@tzurp can you define these properties in a separate config file? I wonder if the wdio config is the best place to do that 🤔

**Reply by tzurp** [reply]
@christian-bromann the thing is that it is working except that the TS compiler is not recognizing the new properties of the config type. I can circumvent this adding `//@ts-ignore`. I prefer this workaround than adding a new mechanism to grab the 'environment' for different runs.

**Reply by christian-bromann** [reply]
With the v7 update we remove the ability to extend the config to avoid cluttering it. My recommendation would be to have an own configuration file for environment data.

**Reply by christian-bromann** [reply]
@tzurp actually it is possible, see Gitter channel:

```ts
declare module NodeJS {
    interface Global {
       mySharedData: any
    }
}
```

**Reply by tzurp** [reply]
> With the v7 update we remove the ability to extend the config to avoid cluttering it.

So it still possible to 'extend' the config using JS right? so it might be weird that TS will not be supported here.
>My recommendation would be to have an own configuration file for environment data.

I suspect that we'll need to use some runner option or another custom config to achieve that?

**Reply by christian-bromann** [reply]
> So it still possible to 'extend' the config using JS right? so it might be weird that TS will not be supported here.

Yeah there is no issue setting custom options, if you use TS there might be type issues which is why you need to extend it

> I suspect that we'll need to use some runner option or another custom config to achieve that?

No, just put your environment config in a file and import it in your test

**Reply by tzurp** [reply]
@christian-bromann thanks!

**Reply by staizen-alexm** [reply]
thread bump! im also looking for this feature. i'm using ts with wdio7 async

i don't fully follow the suggested solution...

can someone help on clarifying this please?

im a  little lost on this part:

> Yeah there is no issue setting custom options, if you use TS there might be type issues which is why you need to extend it

and 

> No, just put your environment config in a file and import it in your test

**Reply by amandaflagg** [reply]
i'm in the same situation as the last commenter. @christian-bromann or @tzurp can you clarify the solution you landed on here? i was trying to store an additional couple configs in the wdio config, but am unable to extend the types, so i always run into type errors with my custom config types.

**Reply by christian-bromann** [reply]
@amandaflagg I don't think it is a good idea to populate the wdio config with data that doesn't configure WebdriverIO. I recommend to create your own config file and import it in your test. Or create a custom command that returns a configuration based on parameters.

**Reply by tzurp** [reply]
@amandaflagg as a 'temporary' solution you can add `// @ts-ignore ` above the line with the error and it would work just fine.

**Reply by rlyonbox** [reply]
The WDIO documentation does list this pattern (https://webdriver.io/docs/browserobject/#custom-configurations) but with no TypeScript support it seems deprecated?
